Research/lva-node-app/public/js/main.js

```diff
@@ -0,0 +1,658 @@
+                        switch (methodName)
+                        {
+                            case 'GraphTopologyDelete':
+                                delete graphTopologies[fullPayload.Payload.name];
+                                break;
+                            case 'GraphInstanceDelete':
+                                delete graphInstances[fullPayload.Payload.name];
+                                break;
+                            case 'GraphInstanceList':
+                                updateGraphsandInstances(methodName, JSON.parse(request.response)[1]);
+                                break;
+                            case 'GraphTopologyList':
+                                updateGraphsandInstances(methodName, JSON.parse(request.response)[1]);
+                                break;
+                            case 'GraphInstanceSet':
+                                graphInstances[fullPayload.Payload.name] = fullPayload;
+                                break;
+                            case 'GraphTopologySet':
+                                // only alert and run display media graphs if request is coming from mediagraphs.html page
+                                graphTopologies[fullPayload.Payload.name] = fullPayload;
+                                if (!isFromOutputPage)
+                                {
+                                    alert("Success!");
+                                    displayMediaGraphs();
+                                }
+                                break;
+                            default:
+                                break;
+                        }
+                        resolve(result);
+                    }
+                }
+                else
+                {
+                    if(request.response != "") alert(request.response);
+                    reject(request.response);
+                }
+            }
+
+        }
+        request.send(JSON.stringify(fullPayload));
+    });
+}
+
+/**
+* stream device to cloud messages from IoT Hub via websocket, opened server side
+*/
+function emitdata()
+{
+    let request = prepareRequest(`http://localhost:${PORT}/hubmessages`, "GET");
+    request.onreadystatechange = function ()
+    {
+        //on successful response. Result is object like [{methodName: 'GraphTopologyList'}, {value: 'long JSON object....'}]
+        if (request.readyState == 4)
+        {
+            if(request.status == 200)
+            {
+                document.getElementById('stop-messages').disabled = false;
+                document.getElementById('start-messages').disabled = true;
+            }
+            else
+            {
+                alert(request.response);
+            }
+        }
+    }
+    request.send();
+}
+
+/**
+ * stop receiving device-cloud IoT Hub messages
+ */
+function stopMessages()
+{
+    let request = prepareRequest(`http://localhost:${PORT}/stopMessages`);
+    request.send();
```

Comment:
this is async, meaning it could complete (or fail) long after stopMessages returns. this seems wrong... we'd basically disable the stop messages button and ignore all errors, even if stop had failed (or not yet completed)

Reply:
promisified